aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -371,6 +371,126 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
   }
 }
 
+#define HANDLE_EQUATION_CASE(MAX1, MAX2, MAX3) \
+  MAX1##idx = n / MAX1##stride; \
+  n = n % MAX1##stride; \
+  MAX2##idx = n / MAX2##stride; \
+  n = n % MAX2##stride; \
+  MAX3##idx = n / MAX3##stride;
+
+template <typename scalar_t, bool upper>
+__global__
+void triu_tril_kernel(
+    scalar_t* result, scalar_t* self, int64_t k, int64_t N,
+    int64_t res_batch_stride, int64_t self_batch_stride,
+    int64_t res_row_stride, int64_t res_col_stride,
+    int64_t max_stride, int64_t min_stride, bool row_is_max_stride) {
+  int64_t linear_idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (linear_idx >= N) {
+    return;
+  }
+
+  int64_t n = linear_idx;
+  int64_t self_batch_idx, max_idx, min_idx;
+
+  // There are three cases:
+  // Case 1: batch_stride >= max_stride > min_stride
+  // Case 2: max_stride >= batch_stride > min_stride
+  // Case 3: max_stride > min_stride >= batch_stride
+  // Since we are dealing with integer indices, a generic way to obtain an integer solution
+  // (x, y, z) for an equation of the form:
+  // n = a.x + b.y + c.z
+  // would be to perform floor division (integer division) in a particular order.
+  // Specifically, we would divide n by the maximum multiplicand to get the variable pertaining
+  // to the maximum multiplicand, and then use the remainder (if any) recursively, to obtain the rest
+  // of the solution.
+  if (self_batch_stride >= max_stride) {  // This is Case 1
+    HANDLE_EQUATION_CASE(self_batch_, max_, min_)
+  } else if (self_batch_stride > min_stride) {  // This is Case 2
+    HANDLE_EQUATION_CASE(max_, self_batch_, min_)
+  } else {  // This is Case 3
+    HANDLE_EQUATION_CASE(max_, min_, self_batch_)
+  }
+
+  int64_t row, col;
+  if (row_is_max_stride) {
+    row = max_idx;
+    col = min_idx;
+  } else {
+    row = min_idx;
+    col = max_idx;
+  }
+
+  bool mask = upper ? (col - row >= k) : (col - row <= k);
+
+  // Now compute the offset for the result tensor
+  int64_t res_offset = self_batch_idx * res_batch_stride + row * res_row_stride + col * res_col_stride;
+  result[res_offset] = mask ? self[linear_idx] : scalar_t(0);
+}
+
+#define DECLARE_TRIL_TRIU_VARIABLES(SELF, RESULT) \
+  int64_t N = SELF.numel(), \
+          self_batch_stride = SELF.dim() > 2 ? SELF.stride(-3) : matrixStride(SELF), \
+          res_batch_stride = RESULT.dim() > 2 ? RESULT.stride(-3) : matrixStride(RESULT), \
+          self_row_stride = SELF.stride(-2), self_col_stride = SELF.stride(-1), \
+          res_row_stride = RESULT.stride(-2), res_col_stride = RESULT.stride(-1); \
+  bool row_is_max_stride = self_row_stride >= self_col_stride; \
+  int64_t max_stride = row_is_max_stride ? self_row_stride : self_col_stride; \
+  int64_t min_stride = row_is_max_stride ? self_col_stride : self_row_stride; \
+  dim3 dim_block = cuda::getApplyBlock(); \
+  dim3 dim_grid; \
+  AT_CHECK(cuda::getApplyGrid(N, dim_grid, SELF.get_device()), "Unable to get dim grid");
```

Comment:
Typical way of handling batched computations is to add grid dimension (dim_grid.y for example) to handle batch elements. That way, you have less index math in the kernel that is good both for readability and for performance.

Reply:
Would the following be appropriate?:
```
dim3 dim_grid((N + dim_block.x - 1) / (dim_block.x * batchCount(SELF)), batchCount(SELF));
```

`batchCount(SELF)` is a function to return the number of matrices batched. This would allow me to use `gridDim.y` as a batch index, as advised.